pkg/image/api/helper.go

```diff
@@ -290,6 +371,30 @@ func DockerImageReferenceForStream(stream *ImageStream) (DockerImageReference, e
 	return ParseDockerImageReference(spec)
 }
 
+// FollowTagReference walks through the defined tags on a stream, following any referential tags in the stream.
+// Will return ok if the tag is valid, multiple if the tag had at least reference, and ref and finalTag will be the last
+// tag seen. If a circular loop is found ok will be false.
+func FollowTagReference(stream *ImageStream, tag string) (finalTag string, ref *TagReference, ok bool, multiple bool) {
```

Comment:
> Not in fewer than three passes, no.

I had trouble too.  See if you can suggest a more understandable iterative (non-recursive) factorization.